### PR TITLE
PERF-#6122: Extract Feather's metadata without reading the whole file

### LINKED ISSUE #6122: `pd.read_feather(file)` actually reads the file twice!
There's a really strange logic in our Feather dispatcher introduced a long time ago that reads the whole Feather file just to extract its column names:
https://github.com/modin-project/modin/blob/632d724adb013d6f4d18aa2a4f2faa9325d324e8/modin/core/io/column_stores/feather_dispatcher.py#L55-L64

Note that `.build_query_compiler(..)` launch parsing kernels that would also be reading the same parquet file but in a distributed way.

We should be using more smarter way to extract feather's metadata: https://issues.apache.org/jira/browse/ARROW-10344

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

This PR rewrites retrieving feather's metadata from reading the whole file to an approach suggested in the related arrow's issue: https://issues.apache.org/jira/browse/ARROW-10344

For the reproducer below, I've measured the following performance on 112 cores Intel(R) Xeon(R) Gold 6238R CPU @ 2.20GHz

| | master | this PR |
|-|-|-|
|read_feather| 6.05s | 0.99s |

```python
import modin.pandas as pd
import numpy as np
import tempfile
from timeit import default_timer as timer

NROWS = 10_000_000
NCOLS = 32

with tempfile.NamedTemporaryFile() as file:
    pd.DataFrame({f"col{i}": np.arange(NROWS) for i in range(NCOLS)}).to_feather(file.name)

    t1 = timer()
    df = pd.read_feather(file.name)
    print("read_feather", timer() - t1)
```

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6122  
- [x] tests are passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/column_stores/feather_dispatcher.py [hunk 1]
```diff
@@ -52,13 +52,15 @@ def _read(cls, path, columns=None, **kwargs):
...
-            from pyarrow.feather import read_feather
+            from pyarrow import ipc
 
             with OpenFile(
                 path,
                 **(kwargs.get("storage_options", None) or {}),
             ) as file:
-                df = read_feather(file)
-            # pyarrow.feather.read_feather doesn't support columns as pandas.Index
-            columns = list(df.columns)
```
Comment (dchigarev) [inline modin/core/io/column_stores/feather_dispatcher.py:63]: `shema.names` already returns a list of columns, so we don't need this explicit conversion anymore

File: modin/core/io/column_stores/feather_dispatcher.py [hunk 2]
```diff
@@ -52,13 +52,15 @@ def _read(cls, path, columns=None, **kwargs):
...
+                # Opens the file to extract its metadata
+                reader = ipc.open_file(file)
+            # TODO: pyarrow's schema contains much more metadata than just column names, it also
+            # has dtypes and index information that we could use when building a dataframe
+            columns = reader.schema.names
         return cls.build_query_compiler(path, columns, use_threads=False)
```
Comment (Garra1980) [inline modin/core/io/column_stores/feather_dispatcher.py:66]: I'm wondering why use_threads is False here...

**Review by vnlitvinov — APPROVED** [review]
LGTM! Though I would like to know answer to @Garra1980 question, too...

**Review by YarShev — APPROVED** [review]

File: modin/core/io/column_stores/feather_dispatcher.py (hunk 2 above)
Comment (YarShev) [reply modin/core/io/column_stores/feather_dispatcher.py:66]: When using a distributed engine like Ray/Dask/Unidist we probably don't want to use threading in remote functions in order not to get into oversubscribing.

**Pull request merged**